File: hu.bme.mit.incqueryd.runtime/hu.bme.mit.incqueryd.core/src/main/java/hu/bme/mit/incqueryd/core/rete/nodes/InputNode.java
/*******************************************************************************
 * Copyright (c) 2010-2014, Gabor Szarnyas, Istvan Rath and Daniel Varro
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 * Gabor Szarnyas - initial API and implementation
 *******************************************************************************/
package hu.bme.mit.incqueryd.core.rete.nodes;

import hu.bme.mit.incqueryd.core.rete.dataunits.ChangeSet;
import hu.bme.mit.incqueryd.core.rete.dataunits.Tuple;

import java.util.HashSet;
import java.util.Set;

import org.eclipse.incquery.runtime.rete.recipes.TypeInputRecipe;

import com.carrotsearch.sizeof.RamUsageEstimator;

public class InputNode implements AlphaNode {

	/**
	 * Input node implementation.
	 */
	InputNode(final TypeInputRecipe recipe) {
		super();
	}

	protected final Set<Tuple> tuples = new HashSet<>();

	public double getMemoryConsumption() {
		return tuples.size() > 0 ? (double) RamUsageEstimator.sizeOf(tuples) / RamUsageEstimator.ONE_MB : 0;
	}

	public Set<Tuple> getTuples() {
		return tuples;
	}
	
	@Override
	public ChangeSet update(final ChangeSet changeSet) {
		final Set<Tuple> incomingTuples = changeSet.getTuples();
		switch (changeSet.getChangeType()) {
		case POSITIVE:
			tuples.addAll(incomingTuples);
			break;
		case NEGATIVE:
			tuples.removeAll(incomingTuples);
			break;
		default:
			break;
		}

		return changeSet;
	}

}

File: hu.bme.mit.incqueryd.core/hu.bme.mit.incqueryd.engine/src/main/scala/hu/bme/mit/incqueryd/engine/rete/nodes/InputNode.scala
package hu.bme.mit.incqueryd.engine.rete.nodes

import org.eclipse.incquery.runtime.rete.recipes.TypeInputRecipe
import java.util.HashSet
import hu.bme.mit.incqueryd.engine.rete.dataunits.Tuple
import java.lang.management.ManagementFactory
import hu.bme.mit.incqueryd.engine.rete.dataunits.ChangeSet
import hu.bme.mit.incqueryd.engine.rete.dataunits.ChangeType
import org.eclipse.incquery.runtime.rete.recipes.BinaryInputRecipe
import org.eclipse.incquery.runtime.rete.recipes.UnaryInputRecipe
import hu.bme.mit.bigmodel.rdf.RDFHelper
import hu.bme.mit.bigmodel.fourstore.FourStoreDriverTrainBenchmark
import scala.collection.JavaConversions.asScalaBuffer
import scala.collection.JavaConversions.asScalaSet
import scala.collection.JavaConversions.collectionAsScalaIterable
import scala.collection.JavaConversions.mapAsScalaMap
import scala.collection.JavaConversions.seqAsJavaList
import scala.collection.immutable.Stack
import hu.bme.mit.incqueryd.engine.rete.messages.UpdateMessage
import hu.bme.mit.incqueryd.engine.rete.dataunits.ReteNodeSlot

class InputNode(recipe: TypeInputRecipe) extends AlphaNode {

  val typeName = recipe.getTypeName().split("#")(1)
  val tuples = new HashSet[Tuple]

  val databaseDriver = new FourStoreDriverTrainBenchmark("trainbenchmark_cluster", false)

  def getMemoryConsumption() = {
    val memoryMXBean = ManagementFactory.getMemoryMXBean();
    memoryMXBean.getHeapMemoryUsage().getUsed() / (1024 * 1024);
  }

  def update(changeSet: ChangeSet) = {
    val incomingTuples = changeSet.getTuples
    changeSet.getChangeType() match {
      case ChangeType.POSITIVE => tuples.addAll(incomingTuples)
      case ChangeType.NEGATIVE => tuples.removeAll(incomingTuples)
    }
    changeSet
  }

  def initialize() = {
    recipe match {
      case binaryInputRecipe: BinaryInputRecipe => {
        val traceInfo = binaryInputRecipe.getTraceInfo
        if (traceInfo.startsWith("attribute")) {
          initializeAttribute(binaryInputRecipe)
        } else if (traceInfo.startsWith("edge")) {
          initializeEdge(binaryInputRecipe)
        }
      }
      case unaryInputRecipe: UnaryInputRecipe => {
        initializeVertex(unaryInputRecipe)
      }
    }

    val changeSet = new ChangeSet(tuples, ChangeType.POSITIVE)
    changeSet
  }

  def initializeAttribute(recipe: BinaryInputRecipe) = {
    val typeName = RDFHelper.brackets(recipe.getTypeName)
    val attributes = databaseDriver.collectVerticesWithProperty(typeName)

    attributes.foreach(attribute => {
      val key = attribute._1
      val value = attribute._2

      val regex = "\"(.*?)\"\\^\\^<http://www.w3.org/2001/XMLSchema#int>".r
      val intValue = regex.findFirstMatchIn(value).get.group(1)
      tuples += new Tuple(key, intValue)
    })
  }

  def initializeEdge(recipe: BinaryInputRecipe) = {
    val typeName = RDFHelper.brackets(recipe.getTypeName)
    val edges = databaseDriver.collectEdges(typeName)

    edges.entries.foreach(edge => {
      tuples += new Tuple(edge.getKey, edge.getValue)
    })
  }

  def initializeVertex(recipe: UnaryInputRecipe) = {
    val vertices = databaseDriver.collectVertices(RDFHelper.brackets(recipe.getTypeName))
    vertices.foreach(vertex => tuples += new Tuple(vertex))
  }
}
